rllib/algorithms/tests/test_algorithm_config.py

```diff
@@ -185,6 +189,225 @@ class A:
+                module_specs={
+                    "p1": SingleAgentRLModuleSpec(module_class=CustomRLModule1),
+                    "p2": SingleAgentRLModuleSpec(module_class=CustomRLModule1),
+                }
+            ),
+        )
+        config.validate()
+
+        spec, expected = self._get_expected_marl_spec(config, CustomRLModule1)
+        self._assertEqualMARLSpecs(spec, expected)
+
+        ########################################
+        # We want to override the default rl module
+        config = SingleAgentAlgoConfig().rl_module(
+            _enable_rl_module_api=True,
+            rl_module_spec=SingleAgentRLModuleSpec(module_class=CustomRLModule1),
+        )
+        config.validate()
+
+        spec, expected = self._get_expected_marl_spec(config, CustomRLModule1)
+        self._assertEqualMARLSpecs(spec, expected)
+
+        # expected module should become the passed module if we pass it in.
+        spec, expected = self._get_expected_marl_spec(
+            config, CustomRLModule2, passed_module_class=CustomRLModule2
+        )
+        self._assertEqualMARLSpecs(spec, expected)
+        ########################################
+        # We now want to tell the algorithm to use a different rl module for each
+        # RLModule
+        config = SingleAgentAlgoConfig().rl_module(
+            _enable_rl_module_api=True,
+            rl_module_spec=MultiAgentRLModuleSpec(
+                module_specs=SingleAgentRLModuleSpec(module_class=CustomRLModule1)
+            ),
+        )
+        config.validate()
+
+        spec, expected = self._get_expected_marl_spec(config, CustomRLModule1)
+        self._assertEqualMARLSpecs(spec, expected)
+
+        # expected module should become the passed module if we pass it in.
+        spec, expected = self._get_expected_marl_spec(
+            config, CustomRLModule2, passed_module_class=CustomRLModule2
+        )
+        self._assertEqualMARLSpecs(spec, expected)
+
+        ########################################
+        # We now want to tell the algorithm to use a heterogeneous rl module for each
+        # RLModule
+        config = SingleAgentAlgoConfig().rl_module(
+            _enable_rl_module_api=True,
+            rl_module_spec=MultiAgentRLModuleSpec(
+                marl_module_class=CustomMARLModule1,
+                module_specs={
+                    "p1": SingleAgentRLModuleSpec(module_class=CustomRLModule1),
+                    "p2": SingleAgentRLModuleSpec(module_class=CustomRLModule1),
+                },
+            ),
+        )
+        config.validate()
+
+        spec, expected = self._get_expected_marl_spec(
+            config, CustomRLModule1, expected_marl_module_class=CustomMARLModule1
+        )
+        self._assertEqualMARLSpecs(spec, expected)
+
+        # this is expected to return CustomRLModule1 instead of CustomRLModule3 which
+        # is passed in. Because the default for p1, p2 is to use CustomRLModule1
```

Comment:
We didn't cover this case this morning and this is not what I'd expect. 
I'd expect the default SingleAgentRLModuleSpecs to be overwritten by `assed_module_class=CustomRLModule3`

Reply:
We have actually not talked about the passed_module_class at all in none of the cases. Added this description. Hope it helps. This is for the multi-agent case during add_policy. it's a bit different. 

```
        # This is expected to return CustomRLModule1 instead of CustomRLModule3 which
        # is passed in. Because the default for p1, p2 is to use CustomRLModule1. The 
        # passed module_spec only sets a default to fall back onto in case the 
        # module_id is not specified in the original MultiAgentRLModuleSpec. Since P1 
        # and P2 are both assigned to CustomeRLModule1, the passed module_spec will not 
        # be used. This is the expected behavior for adding a new modules to a 
        # multi-agent RLModule that is not defined in the original 
        # MultiAgentRLModuleSpec.
```